docs/rfcs/0002/readme.md

```diff
@@ -37,21 +34,58 @@ The proposal below might take a couple of read-throughs; I've also added a concr
-4. The USER who desires to use the application works with the AUTHOR and the related NNF Container Profile to understand the storage requirements.
-5. The USER submits a WLM job with the #DW container fields populated
-6. WLM runs the job and drives the job through the following stages...
-    1. Proposal: RABBIT validates the #DW container directive by comparing the supplied values to what is listed in the NNF Container Profile. If the USER fails to meet the requirements, the job fails. 
-    2. Pre-run: RABBIT software will:
-        1. create a config map reflecting the storage requirements and any runtime parameters; this is provided to the container at the volume mount named "nnf-config", if specified.
-        2. duplicate the pod template specification from the Container Profile and patches the necessary Volumes and the config map. The spec is used as the basis for starting the necessary pods and containers.
-    3. The containerized application executes. The expected mounts are available per the requirements and celebration occurs.
-
-Example
--------
-
-Say I authored a simple application, `foo`, that requires Rabbit local GFS2 storage and a persistent Lustre storage volume. As the author, my program is coded to expect the GFS2 volume is mounted at `/foo/local` and the Lustre volume is mounted at `/foo/persistent`. In this case, the storages are not optional, so they are defined as such in the NNF Container Profile.
+    3. the NNF storage requirements described above.
+3. The ADMINISTRATOR creates a corresponding _NNF Container Profile_ Kubernetes custom resource with the necessary NNF storage requirements and pod specification as described by the AUTHOR
+4. The USER who desires to use the application works with the AUTHOR and the related NNF Container Profile to understand the storage requirements
+5. The USER submits a WLM job with the #DW container directive variables populated
+6. WLM runs the workflow and drives it through the following stages...
+    1. `Proposal`: RABBIT validates the #DW container directive by comparing the supplied values to those listed in the NNF Container Profile. If the workflow fails to meet the requirements, the job fails
+    2. `PreRun`: RABBIT software:
+        1. creates a config map reflecting the storage requirements and any runtime parameters; this is provided to the container at the volume mount named `nnf-config`, if specified
+        2. duplicates the pod template specification from the Container Profile and patches the necessary Volumes and the config map. The spec is used as the basis for starting the necessary pods and containers.
+    3. The containerized application(s) executes. The expected mounts are available per the requirements and celebration occurs. The pods continue to run until:
+       1. the pod completes successfully (any failed pods will be restarted)
+       2. timeout (i.e. `activeDeadlineSeconds`) is hit (optional)
```

Comment:
This doesn't make sense to me - How does a developer know how long the compute application will run? I think `activeDeadlineSeconds` is only useful _after_ transitioning to PostRun. But that would require us to change the value at runtime - can we do this?

Reply:
This is contraint for the k8s Job. If you set this to 300, it will kill the job's pods after 300 seconds and consider it failed.

Reply:
Say I'm authoring SCR. My application runs alongside the compute resources however long the computes run (Flux decides). At that stage, why would you ever set a deadline in SCR? It's completely beholden to the compute application (and therefore Flux).

After Flux transitions the workflow to PostRun; that's when the timeout should start. But I'm not sure we can change the deadline after creation.

Reply:
That makes more sense. In this case, I was exposing k8s Job's deadline configuration. Maybe that's not helpful since they won't care how long the containers are running while in PreRun. Just PostRun.

When they transition to PostRun, it will stay `Ready:false` until the jobs are done running. At that point, Rabbit checks the return values of all the jobs to see if they passed. If they did, it will transition to `Ready:true` only if all the jobs are done and if they all succeeded.

If we start a timeout at the transition to PostRun, and pods are still running, what is the Rabbit SW expected to do here? Kill the pods? Flux could do that by transitioning to teardown if in the PostRun stage for X number of seconds.

Reply:
> If we start a timeout at the transition to PostRun, and pods are still running, what is the Rabbit SW expected to do here? 
What does Kubernetes do to a Job exceeding the deadline?

I would expect the workflow to fail if that case

> Flux could do that by transitioning to teardown if in the PostRun stage for X number of seconds.
Flux will always do that - be it a stuck workflow or whatever. But as you have it presented `activeDeadlineSeconds` is a value the application author would decide on. So this is our opportunity to provide a specific error code / cause / reason to Flux that is valuable to the parties involved: Flux, User, and Author

Reply:
> What does Kubernetes do to a Job exceeding the deadline? I would expect Rabbit S/W to fail the workflow.

It will kill the pods and consider the Job failed; this deadline starts when the pods start. Like you pointed out, that's not helpful because the author doesn't know how long the workflow will be in PreRun (and probably doesn't care). I didn't fully understand that when I wrote this part up.

So I don't think we should even use this deadline, and we'll need to implement our own timer that starts in PostRun if that timer is hit, we can kill the pods that aren't finished.

> So this is our opportunity to provide a specific error code / cause / reason to Flux that is valuable to the parties involved: Flux, User, and Author

Do we set the workflow to have an error status then? (i.e. `Ready:false` & `Status:Error`)

Reply:
> Do we set the workflow to have an error status then? (i.e. Ready:false & Status:Error)

Yes. Also, a Message field that's meant to convene some helpful information.